llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp

```diff
@@ -3643,6 +3645,54 @@ Instruction *InstCombinerImpl::visitOr(BinaryOperator &I) {
             foldAddLikeCommutative(I.getOperand(1), I.getOperand(0),
                                    /*NSW=*/true, /*NUW=*/true))
       return R;
+
+    Value *Cond0 = nullptr, *Cond1 = nullptr;
+    const APInt *Op0Eq = nullptr, *Op0Ne = nullptr;
+    const APInt *Op1Eq = nullptr, *Op1Ne = nullptr;
+
+    //  (!(A & N) ? 0 : N * C) + (!(A & M) ? 0 : M * C) -> A & (N + M) * C
+    if (match(I.getOperand(0),
+              m_Select(m_Value(Cond0), m_APInt(Op0Eq), m_APInt(Op0Ne))) &&
+        match(I.getOperand(1),
+              m_Select(m_Value(Cond1), m_APInt(Op1Eq), m_APInt(Op1Ne)))) {
+      CmpPredicate Pred0, Pred1;
+
+      auto LHSDecompose =
+          decomposeBitTest(Cond0, /*LookThruTrunc=*/true,
```

Comment:
@andjo403 Do you think it is overkill to use `decomposeBitTest`? The only thing I want to match are `icmp ne/eq (and X, Power2), 0` and `trunc X to i1`. It seems that the mask of `decomposeBitTest(relational icmp)` is always non-power-of-2 :(

Reply:
to me it feels better to have one common pattern by calling decomposeBitTest.